File: src/backends/native/owned-aggregate-leases.mjs
/**
 * Private native acquisition ledger for resource-bearing aggregate conversions.
 * Host projections keep these contexts, scopes and batches out of public values.
 *
 * @file
 */
export const ownedAggregateLeaseSource = `
#include <lean/lean.h>
#include "lean_bridge_native_runtime.h"
#include <pthread.h>
#include <stdatomic.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifndef LB_OWNED_ALLOC
#define LB_OWNED_ALLOC(size) calloc(1, (size))
#endif
#ifndef LB_OWNED_FREE
#define LB_OWNED_FREE(pointer) free(pointer)
#endif

enum {
  LB_OWNED_OK = 0, LB_OWNED_INVALID = 1, LB_OWNED_LIMIT = 2,
  LB_OWNED_ALLOC_FAILED = 3, LB_OWNED_CLOSED = 4,
  LB_OWNED_THREAD = 5, LB_OWNED_PROCESS = 6, LB_OWNED_RUNTIME = 7,
  LB_OWNED_ORDER = 8, LB_OWNED_RETAINED_LIMIT = 4096,
  LB_OWNED_SCOPE_LIMIT = 64
};

typedef struct lb_owned_context lb_owned_context;
typedef struct lb_owned_scope lb_owned_scope;
typedef struct lb_owned_batch lb_owned_batch;
typedef struct lb_owned_owner {
  struct lb_owned_owner *next;
  const char *kind;
  lean_object *value;
  uint64_t token;
  size_t references;
} lb_owned_owner;
typedef struct lb_owned_entry {
  struct lb_owned_entry *next;
  lb_owned_owner *owner;
} lb_owned_entry;
struct lb_owned_batch {
  lb_owned_batch *next;
  lb_owned_context *context;
  lb_owned_entry *entries;
};
struct lb_owned_scope {
  lb_owned_context *context;
  lb_owned_scope *parent;
  lb_owned_entry *inputs;
  lb_owned_entry *outputs;
  size_t retained;
};
struct lb_owned_context {
  lb_owned_context *self;
  const char *component;
  pid_t process;
  pthread_t thread;
  uint64_t thread_serial;
  uint64_t runtime;
  uint64_t domain;
  lb_owned_owner *owners;
  lb_owned_batch *batches;
  lb_owned_scope *top;
  size_t live_owners;
  size_t scopes;
  int initialized;
  int closing;
  int poisoned;
};

/* A pthread_t can be reused after exit. Serials never wrap or repeat. */
static _Atomic uint64_t lb_owned_next_thread;
static _Thread_local uint64_t lb_owned_thread_serial;
static int lb_owned_identify_thread(void) {
  if (lb_owned_thread_serial) return LB_OWNED_OK;
  uint64_t previous = atomic_load(&lb_owned_next_thread);
  do {
    if (previous == UINT64_MAX) return LB_OWNED_LIMIT;
  } while (!atomic_compare_exchange_weak(&lb_owned_next_thread, &previous, previous + 1));
  lb_owned_thread_serial = previous + 1;
  return LB_OWNED_OK;
}

/* Context and batch storage belongs to the generated host adapter. Component
   and nominal-kind strings are immutable generated constants. A batch may not
   move while registered. Context close defers until the last active scope ends. */
static int lb_owned_affinity(const lb_owned_context *context) {
  if (!context || !context->initialized) return LB_OWNED_CLOSED;
  if (context->self != context) return LB_OWNED_INVALID;
  /* Check the process before entering a mutex inherited across fork. */
  if (context->process != getpid()) return LB_OWNED_PROCESS;
  if (!pthread_equal(context->thread, pthread_self()) || context->thread_serial != lb_owned_thread_serial)
    return LB_OWNED_THREAD;
  return LB_OWNED_OK;
}
static int lb_owned_ready(const lb_owned_context *context) {
  int status = lb_owned_affinity(context);
  if (status) return status;
  if (context->closing || context->poisoned) return LB_OWNED_CLOSED;
  if (!lean_bridge_native_component_ready(context->component)) return LB_OWNED_RUNTIME;
  lean_bridge_native_snapshot snapshot;
  lean_bridge_native_snapshot_read(&snapshot);
  if (snapshot.runtime_instance_id != context->runtime || snapshot.identity_domain_id != context->domain)
    return LB_OWNED_RUNTIME;
  return LB_OWNED_OK;
}
static int lb_owned_context_init(lb_owned_context *context, const char *component) {
  if (!context || !component || !*component) return LB_OWNED_INVALID;
  /* The caller supplies fresh, zero-initialized storage, never a live context. */
  if (context->self || context->initialized || context->owners || context->batches || context->top)
    return LB_OWNED_INVALID;
  int status = lb_owned_identify_thread();
  if (status) return status;
  if (!lean_bridge_native_component_ready(component)) return LB_OWNED_RUNTIME;
  lean_bridge_native_snapshot snapshot;
  lean_bridge_native_snapshot_read(&snapshot);
  *context = (lb_owned_context){0};
  context->self = context;
  context->component = component; context->process = getpid(); context->thread = pthread_self();
  context->thread_serial = lb_owned_thread_serial;
  context->runtime = snapshot.runtime_instance_id; context->domain = snapshot.identity_domain_id;
  context->initialized = 1;
  return LB_OWNED_OK;
}
static lb_owned_owner *lb_owned_find(lb_owned_context *context, const char *kind, uint64_t token) {
  for (lb_owned_owner *owner = context->owners; owner; owner = owner->next)
    if (owner->token == token && strcmp(owner->kind, kind) == 0) return owner;
  return NULL;
}
static void lb_owned_drop(lb_owned_context *context, lb_owned_owner *owner) {
  if (--owner->references) return;
  lb_owned_owner **position = &context->owners;
  while (*position != owner) position = &(*position)->next;
  *position = owner->next;
  /* Other contexts or adapters can still own the same broker identity. */
  if (lean_bridge_native_identity_release(owner->token, owner->kind, owner->value) < 0)
    context->poisoned = 1;
  lean_dec(owner->value); --context->live_owners; LB_OWNED_FREE(owner);
}
static void lb_owned_drop_entries(lb_owned_context *context, lb_owned_entry *entry) {
  while (entry) {
    lb_owned_entry *next = entry->next;
    lb_owned_drop(context, entry->owner); LB_OWNED_FREE(entry); entry = next;
  }
}
static void lb_owned_finish_close(lb_owned_context *context) {
  if (!context->closing || context->top) return;
  while (context->batches) {
    lb_owned_batch *batch = context->batches;
    context->batches = batch->next;
    lb_owned_drop_entries(context, batch->entries);
    *batch = (lb_owned_batch){0};
  }
  context->initialized = 0;
}
static int lb_owned_context_close(lb_owned_context *context) {
  int status = lb_owned_affinity(context);
  if (status) return status;
  context->closing = 1; lb_owned_finish_close(context);
  return LB_OWNED_OK;
}
static int lb_owned_batch_release(lb_owned_context *context, lb_owned_batch *batch) {
  int status = lb_owned_affinity(context);
  if (status) return status;
  lb_owned_batch **position = &context->batches;
  while (*position && *position != batch) position = &(*position)->next;
  /* Check membership before dereferencing an unregistered or already closed batch. */
  if (!*position) return LB_OWNED_INVALID;
  *position = batch->next;
  lb_owned_drop_entries(context, batch->entries); *batch = (lb_owned_batch){0};
  return context->poisoned ? LB_OWNED_RUNTIME : LB_OWNED_OK;
}
static int lb_owned_scope_begin(lb_owned_context *context, lb_owned_scope *scope) {
  int status = lb_owned_ready(context);
  if (status) return status;
  if (!scope || scope->context || scope->parent || scope->inputs || scope->outputs || scope->retained)
    return LB_OWNED_INVALID;
  for (lb_owned_scope *active = context->top; active; active = active->parent)
    if (scope == active) return LB_OWNED_INVALID;
  if (context->scopes == LB_OWNED_SCOPE_LIMIT) return LB_OWNED_LIMIT;
  *scope = (lb_owned_scope){0}; scope->context = context; scope->parent = context->top;
  context->top = scope; ++context->scopes;
  return LB_OWNED_OK;
}
static int lb_owned_scope_ready(lb_owned_scope *scope) {
  if (!scope || !scope->context) return LB_OWNED_INVALID;
  int status = lb_owned_ready(scope->context);
  if (status) return status;
  return scope->context->top == scope ? LB_OWNED_OK : LB_OWNED_ORDER;
}
static int lb_owned_scope_hold(lb_owned_scope *scope, const char *kind, lean_object *value,
    lb_owned_owner *existing, int input, uint64_t *token) {
  int status = lb_owned_scope_ready(scope);
  if (status) return status;
  if (!kind || !*kind || !value || lean_is_scalar(value) || !token) return LB_OWNED_INVALID;
  lb_owned_context *context = scope->context;
  if (scope->retained == LB_OWNED_RETAINED_LIMIT) return LB_OWNED_LIMIT;
  lb_owned_owner *owner = existing;
  if (!owner) for (lb_owned_owner *item = context->owners; item; item = item->next)
    if (item->value == value && strcmp(item->kind, kind) == 0) { owner = item; break; }
  if (owner && owner->references == SIZE_MAX) return LB_OWNED_LIMIT;
  if (!owner && context->live_owners == LB_OWNED_RETAINED_LIMIT) return LB_OWNED_LIMIT;
  lb_owned_entry *entry = (lb_owned_entry *)LB_OWNED_ALLOC(sizeof(*entry));
  if (!entry) return LB_OWNED_ALLOC_FAILED;
  if (!owner) {
    owner = (lb_owned_owner *)LB_OWNED_ALLOC(sizeof(*owner));
    if (!owner) { LB_OWNED_FREE(entry); return LB_OWNED_ALLOC_FAILED; }
    *owner = (lb_owned_owner){0};
    owner->token = lean_bridge_native_identity_acquire(kind, value);
    if (!owner->token) { LB_OWNED_FREE(owner); LB_OWNED_FREE(entry); return LB_OWNED_LIMIT; }
    owner->kind = kind; owner->value = value; lean_inc(value);
    owner->next = context->owners; context->owners = owner; ++context->live_owners;
  }
  ++owner->references; ++scope->retained; entry->owner = owner;
  lb_owned_entry **head = input ? &scope->inputs : &scope->outputs;
  entry->next = *head; *head = entry; *token = owner->token;
  return LB_OWNED_OK;
}
static int lb_owned_scope_acquire(lb_owned_scope *scope, const char *kind, lean_object *value, uint64_t *token) {
  return lb_owned_scope_hold(scope, kind, value, NULL, 0, token);
}
static int lb_owned_scope_borrow(lb_owned_scope *scope, const char *kind, uint64_t token, lean_object **value) {
  int status = lb_owned_scope_ready(scope);
  if (status) return status;
  if (!kind || !*kind || !token || !value) return LB_OWNED_INVALID;
  lb_owned_owner *owner = lb_owned_find(scope->context, kind, token);
  if (!owner) return LB_OWNED_INVALID;
  uint64_t retained = 0;
  status = lb_owned_scope_hold(scope, kind, owner->value, owner, 1, &retained);
  if (!status) *value = owner->value;
  return status;
}
static int lb_owned_scope_abort(lb_owned_scope *scope) {
  if (!scope || !scope->context) return LB_OWNED_INVALID;
  lb_owned_context *context = scope->context;
  int status = lb_owned_affinity(context);
  if (status) return status;
  if (context->top != scope) return LB_OWNED_ORDER;
  lb_owned_drop_entries(context, scope->outputs); lb_owned_drop_entries(context, scope->inputs);
  context->top = scope->parent; --context->scopes; *scope = (lb_owned_scope){0};
  lb_owned_finish_close(context);
  return context->poisoned ? LB_OWNED_RUNTIME : LB_OWNED_OK;
}
static int lb_owned_scope_commit(lb_owned_scope *scope, lb_owned_batch *batch) {
  int status = lb_owned_scope_ready(scope);
  if (status) return status; /* Caller must abort any failed commit. */
  if (!batch) return LB_OWNED_INVALID;
  lb_owned_context *context = scope->context;
  for (lb_owned_batch *active = context->batches; active; active = active->next)
    if (active == batch) return LB_OWNED_INVALID;
  /* Caller supplies zero-initialized output storage, never another context's batch. */
  if (batch->context || batch->entries || batch->next) return LB_OWNED_INVALID;
  batch->context = context; batch->entries = scope->outputs; batch->next = context->batches;
  context->batches = batch; scope->outputs = NULL;
  return lb_owned_scope_abort(scope);
}
`;

File: tests/fixtures/structured-types/owned-aggregate-leases.c
/* Included after the independent carrier oracle in one native translation unit. */
#include <sys/wait.h>

static size_t live_allocations, allocation_attempts, fail_at, allocation_failures;
static void *ledger_allocate(size_t size) {
  if (++allocation_attempts == fail_at) return NULL;
  void *value = calloc(1, size); if (value) ++live_allocations; return value;
}
static void ledger_free(void *value) {
  if (value) { CHECK(live_allocations > 0); --live_allocations; free(value); }
}
#define LB_OWNED_ALLOC ledger_allocate
#define LB_OWNED_FREE ledger_free
#include "owned-leases.h"

static const char ticket_kind[] = "resource:Owned.Ticket";
static void clean(void) {
  lean_bridge_native_snapshot snapshot; lean_bridge_native_snapshot_read(&snapshot);
  CHECK(snapshot.live_identities == 0); CHECK(live_allocations == 0);
}
static void test_ledger(void) {
  lean_object *value = ticket(42), *borrowed = NULL;
  lb_owned_context context = {0}, peer = {0};
  lb_owned_scope scope = {0}, other = {0};
  lb_owned_batch batch = {0}, output = {0};
  uint64_t token = 0, duplicate = 0, stale = 0;
  CHECK(lb_owned_context_init(&context, "not-initialized") == LB_OWNED_RUNTIME);
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_INVALID);
  CHECK(lb_owned_context_init(&peer, COMPONENT_ID) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&peer, &scope) == LB_OWNED_INVALID);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(!lean_is_exclusive(value));
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &duplicate) == LB_OWNED_OK);
  CHECK(token != 0 && token == duplicate && context.live_owners == 1);
  lb_owned_context copied_context = context;
  CHECK(lb_owned_context_close(&copied_context) == LB_OWNED_INVALID);
  CHECK(lb_owned_scope_begin(&copied_context, &other) == LB_OWNED_INVALID);
  CHECK(lb_owned_scope_borrow(&scope, "wrong nominal type", token, &borrowed) == LB_OWNED_INVALID);
  CHECK(borrowed == NULL);
  CHECK(lb_owned_scope_borrow(&scope, ticket_kind, token, &borrowed) == LB_OWNED_OK);
  CHECK(borrowed == value);
  CHECK(lb_owned_scope_commit(&scope, &batch) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_borrow(&scope, ticket_kind, token, &borrowed) == LB_OWNED_OK);
  CHECK(lb_owned_batch_release(&context, &batch) == LB_OWNED_OK);
  CHECK(lb_owned_batch_release(&context, &batch) == LB_OWNED_INVALID);
  CHECK(!lean_is_exclusive(value)); /* The in-flight input pin still owns it. */
  expect_ticket(F_retainTicket(carry(share(borrowed))), value, 42);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, borrowed, &duplicate) == LB_OWNED_OK);
  CHECK(duplicate == token);
  CHECK(lb_owned_scope_commit(&scope, &output) == LB_OWNED_OK);
  lb_owned_batch copied = output;
  CHECK(lb_owned_batch_release(&context, &copied) == LB_OWNED_INVALID);
  CHECK(lb_owned_batch_release(&peer, &output) == LB_OWNED_INVALID);
  CHECK(lb_owned_batch_release(&context, &output) == LB_OWNED_OK);
  CHECK(lean_is_exclusive(value));
  stale = token;
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  borrowed = NULL;
  CHECK(lb_owned_scope_borrow(&scope, ticket_kind, stale, &borrowed) == LB_OWNED_INVALID);
  CHECK(borrowed == NULL);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(token != stale);
  CHECK(lb_owned_scope_commit(&scope, &batch) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&peer, &other) == LB_OWNED_OK);
  CHECK(lb_owned_scope_borrow(&other, ticket_kind, token, &borrowed) == LB_OWNED_INVALID);
  CHECK(lb_owned_scope_acquire(&other, ticket_kind, value, &duplicate) == LB_OWNED_OK);
  CHECK(duplicate == token); /* One runtime identity, two independent strong owners. */
  CHECK(lb_owned_scope_commit(&other, &output) == LB_OWNED_OK);
  CHECK(lb_owned_batch_release(&context, &batch) == LB_OWNED_OK);
  CHECK(context.poisoned == 0 && !lean_is_exclusive(value));
  CHECK(lb_owned_batch_release(&peer, &output) == LB_OWNED_OK);
  CHECK(lean_is_exclusive(value));
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &other) == LB_OWNED_OK);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_ORDER);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_ORDER);
  CHECK(lb_owned_scope_acquire(&other, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(lb_owned_scope_abort(&other) == LB_OWNED_OK);
  CHECK(lb_owned_scope_commit(&scope, NULL) == LB_OWNED_INVALID);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_INVALID);
  CHECK(lean_is_exclusive(value));
  CHECK(lb_owned_context_close(&peer) == LB_OWNED_OK);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_CLOSED);
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_INVALID);
  lean_dec(value); clean();
}

static size_t fault_case(size_t failure) {
  clean(); allocation_attempts = 0; fail_at = failure;
  lb_owned_context context = {0}; lb_owned_scope scope = {0}; lb_owned_batch batch = {0};
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  int status = LB_OWNED_OK;
  for (unsigned i = 0; i < 24 && !status; ++i) {
    lean_object *value = ticket(i), *borrowed = NULL;
    uint64_t token = UINT64_MAX, duplicate = 0;
    status = lb_owned_scope_acquire(&scope, ticket_kind, value, &token);
    if (status) CHECK(token == UINT64_MAX);
    if (!status) status = lb_owned_scope_acquire(&scope, ticket_kind, value, &duplicate);
    if (!status) { CHECK(duplicate == token); status = lb_owned_scope_borrow(&scope, ticket_kind, token, &borrowed); }
    if (!status) CHECK(borrowed == value);
    lean_dec(value);
  }
  if (failure) {
    CHECK(status == LB_OWNED_ALLOC_FAILED); ++allocation_failures;
    CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_OK);
  } else {
    CHECK(status == LB_OWNED_OK);
    CHECK(lb_owned_scope_commit(&scope, &batch) == LB_OWNED_OK);
    CHECK(lb_owned_batch_release(&context, &batch) == LB_OWNED_OK);
  }
  CHECK(context.live_owners == 0 && context.scopes == 0);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK);
  clean(); fail_at = 0; return allocation_attempts;
}

static void test_limits(void) {
  lb_owned_context context = {0}; lb_owned_scope scopes[LB_OWNED_SCOPE_LIMIT + 1] = {{0}};
  lean_object *value = ticket(7); uint64_t token = 0;
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  for (size_t i = 0; i < LB_OWNED_SCOPE_LIMIT; ++i) CHECK(lb_owned_scope_begin(&context, &scopes[i]) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scopes[LB_OWNED_SCOPE_LIMIT]) == LB_OWNED_LIMIT);
  for (size_t i = LB_OWNED_SCOPE_LIMIT; i > 0; --i) CHECK(lb_owned_scope_abort(&scopes[i - 1]) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scopes[0]) == LB_OWNED_OK);
  for (size_t i = 0; i < LB_OWNED_RETAINED_LIMIT; ++i)
    CHECK(lb_owned_scope_acquire(&scopes[0], ticket_kind, value, &token) == LB_OWNED_OK);
  token = UINT64_MAX;
  CHECK(lb_owned_scope_acquire(&scopes[0], ticket_kind, value, &token) == LB_OWNED_LIMIT);
  CHECK(token == UINT64_MAX && context.live_owners == 1);
  CHECK(lb_owned_scope_abort(&scopes[0]) == LB_OWNED_OK);
  CHECK(lean_is_exclusive(value)); lean_dec(value);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK); clean();
}

static void test_capture_lifetime(void) {
  lb_owned_context context = {0}; lb_owned_scope scope = {0}; lb_owned_batch parent = {0}, child = {0};
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  lean_object *item = ticket(42), *aggregate = bundle(item), *value = take(share(aggregate));
  uint64_t parent_token = 0, child_token = 0;
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, "aggregate:Owned.Bundle", value, &parent_token) == LB_OWNED_OK);
  CHECK(lb_owned_scope_commit(&scope, &parent) == LB_OWNED_OK);
  lean_dec(value);
  lean_object *closure = F_makeRecord(aggregate);
  CHECK(lb_owned_batch_release(&context, &parent) == LB_OWNED_OK);
  lean_dec(item); /* The closure alone now keeps its captured resources alive. */
  lean_object *supplied = ticket(73);
  aggregate = RECORD_CLOSURE_apply(closure, carry(lean_box(1)), bundle(supplied));
  lean_object *projected = take(F_primary(share(aggregate)));
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, projected, &child_token) == LB_OWNED_OK);
  CHECK(lb_owned_scope_commit(&scope, &child) == LB_OWNED_OK);
  lean_dec(aggregate); lean_dec(projected);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_borrow(&scope, ticket_kind, child_token, &value) == LB_OWNED_OK);
  CHECK(lean_is_exclusive(value)); /* One ledger reference, independent of pins. */
  CHECK(lb_owned_batch_release(&context, &child) == LB_OWNED_OK);
  expect_ticket(carry(share(value)), value, 42);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_OK);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK);
  CHECK(lean_is_exclusive(supplied)); lean_dec(supplied); clean();
}

static void test_broker_capacity(void) {
  uint64_t identities[4096]; unsigned values[4096];
  for (size_t i = 0; i < 4096; ++i) {
    identities[i] = lean_bridge_native_identity_acquire("external-owner", &values[i]);
    CHECK(identities[i] != 0);
  }
  lb_owned_context context = {0}; lb_owned_scope scope = {0};
  lean_object *value = ticket(7); uint64_t token = UINT64_MAX;
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_LIMIT);
  CHECK(token == UINT64_MAX && lean_is_exclusive(value));
  CHECK(live_allocations == 0 && context.live_owners == 0 && scope.retained == 0);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_OK);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK); lean_dec(value);
  for (size_t i = 0; i < 4096; ++i)
    CHECK(lean_bridge_native_identity_release(identities[i], "external-owner", &values[i]) == 1);
  clean();
}

typedef struct { lb_owned_context *context; lb_owned_batch *batch; int status; } affinity_probe;
static void *wrong_thread(void *argument) {
  affinity_probe *probe = argument; lb_owned_scope scope = {0};
  probe->status = lb_owned_scope_begin(probe->context, &scope);
  if (probe->status == LB_OWNED_THREAD) probe->status = lb_owned_batch_release(probe->context, probe->batch);
  return NULL;
}
static void *thread_generation(void *argument) {
  affinity_probe *probe = argument;
  if (!probe->context->initialized) {
    probe->status = lb_owned_context_init(probe->context, COMPONENT_ID);
    return NULL;
  }
  lb_owned_context fresh = {0};
  probe->status = lb_owned_context_init(&fresh, COMPONENT_ID);
  if (probe->status) return NULL;
  /* Force OS identifier reuse regardless of the platform's scheduling policy. */
  probe->context->thread = pthread_self();
  probe->status = lb_owned_ready(probe->context);
  if (fresh.thread_serial == probe->context->thread_serial) probe->status = LB_OWNED_INVALID;
  if (lb_owned_context_close(&fresh) != LB_OWNED_OK) probe->status = LB_OWNED_INVALID;
  return NULL;
}
static void *thread_exhaustion(void *argument) {
  affinity_probe *probe = argument; lb_owned_context context = {0};
  probe->status = lb_owned_context_init(&context, COMPONENT_ID); return NULL;
}
static void test_thread_generations(void) {
  lb_owned_context stale = {0}; affinity_probe probe = { &stale, NULL, -1 }; pthread_t thread;
  CHECK(pthread_create(&thread, NULL, thread_generation, &probe) == 0);
  CHECK(pthread_join(thread, NULL) == 0 && probe.status == LB_OWNED_OK);
  CHECK(pthread_create(&thread, NULL, thread_generation, &probe) == 0);
  CHECK(pthread_join(thread, NULL) == 0 && probe.status == LB_OWNED_THREAD);
  uint64_t saved = atomic_load(&lb_owned_next_thread);
  atomic_store(&lb_owned_next_thread, UINT64_MAX);
  CHECK(pthread_create(&thread, NULL, thread_exhaustion, &probe) == 0);
  CHECK(pthread_join(thread, NULL) == 0 && probe.status == LB_OWNED_LIMIT);
  atomic_store(&lb_owned_next_thread, saved); clean();
}
static void test_affinity_and_close(void) {
  lb_owned_context context = {0}; lb_owned_scope scope = {0}, nested = {0}; lb_owned_batch batch = {0};
  uint64_t token = 0; lean_object *value = ticket(7);
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(lb_owned_scope_commit(&scope, &batch) == LB_OWNED_OK);
  affinity_probe probe = { &context, &batch, 0 }; pthread_t thread;
  CHECK(pthread_create(&thread, NULL, wrong_thread, &probe) == 0);
  CHECK(pthread_join(thread, NULL) == 0 && probe.status == LB_OWNED_THREAD);
  pid_t child = fork(); CHECK(child >= 0);
  if (child == 0) {
    int status = lb_owned_scope_begin(&context, &scope);
    _exit(status == LB_OWNED_PROCESS && lb_owned_batch_release(&context, &batch) == LB_OWNED_PROCESS ? 0 : 1);
  }
  int status = 0; CHECK(waitpid(child, &status, 0) == child && WIFEXITED(status) && WEXITSTATUS(status) == 0);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &nested) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&nested, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK);
  CHECK(context.initialized && !lean_is_exclusive(value));
  CHECK(lb_owned_scope_commit(&nested, &batch) == LB_OWNED_CLOSED);
  CHECK(lb_owned_scope_abort(&nested) == LB_OWNED_OK);
  CHECK(context.initialized);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_OK);
  CHECK(!context.initialized && batch.context == NULL && lean_is_exclusive(value));
  lean_dec(value); clean();
}

static void test_retirement(void) {
  lb_owned_context context = {0}; lb_owned_scope scope = {0}; lb_owned_batch batch = {0};
  lean_object *value = ticket(7); uint64_t token = 0;
  CHECK(lb_owned_context_init(&context, COMPONENT_ID) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_OK);
  CHECK(lb_owned_scope_commit(&scope, &batch) == LB_OWNED_OK);
  CHECK(lb_owned_scope_begin(&context, &scope) == LB_OWNED_OK);
  CHECK(lb_owned_scope_acquire(&scope, ticket_kind, value, &token) == LB_OWNED_OK);
  lean_bridge_native_runtime_retire();
  CHECK(lb_owned_scope_commit(&scope, &batch) == LB_OWNED_RUNTIME);
  CHECK(lb_owned_scope_abort(&scope) == LB_OWNED_OK);
  CHECK(lb_owned_batch_release(&context, &batch) == LB_OWNED_OK);
  CHECK(lean_is_exclusive(value)); lean_dec(value);
  CHECK(lb_owned_context_close(&context) == LB_OWNED_OK); clean();
}

int main(void) {
  CHECK(lean_bridge_native_component_initialize(COMPONENT_ID, initialize_component));
  if (getenv("LEAN_BRIDGE_OWNED_COLD_ONLY")) { puts("{\"cold\":true}"); return 0; }
  test_ledger(); test_carriers(); test_capture_lifetime(); test_limits(); test_affinity_and_close();
  test_broker_capacity(); test_thread_generations();
  size_t attempts = fault_case(0);
  for (size_t failure = 1; failure <= attempts; ++failure) fault_case(failure);
  CHECK(allocation_failures == attempts);
  test_retirement();
  lean_bridge_native_snapshot snapshot; lean_bridge_native_snapshot_read(&snapshot);
  printf("{\"checks\":%u,\"exports\":22,\"depth\":128,\"allocationFailures\":%zu,\"liveAllocations\":%zu,\"liveIdentities\":%u}\n",
    checks, allocation_failures, live_allocations, snapshot.live_identities);
  return 0;
}
